proposals/0127-get-sysvar-syscall.md

```diff
@@ -0,0 +1,203 @@
+[SIMD-0088](./0088-enable-core-bpf-programs.md) will require at least five new
+syscalls if the existing design is not ratified.
+
+This tightly-coupled relationship between sysvars and syscalls will perpetually
+swell the syscall interface as changes are made, making maintainence and
+reimplementation difficult.
+
+## Alternatives Considered
+
+Other mechanisms considered that provide sysvar data to BPF programs are as
+follows:
+
+### Sysvar-Specific Syscalls
+
+This approach would involve changing nothing. We'd continue on the same course,
+directly coupling new sysvar fetch functions to new syscalls.
+
+### New VM Memory Regions
+
+A new memory region in the VM could be introduced to store sysvar data, which
+could then be made available to BPF programs through an interface around raw
+pointers, which loads the data similarly to the interface proposed herein.
+
+This approach would depend on direct mapping of host memory to VM memory,
+which has been in development for some time and requires further research
+and development to introduce safely to the VM ABI. This extensive R&D period
+would mean Core BPF program initiatives would remain blocked.
+
+Without direct mapping, sysvar data would be fully copied to this memory
+region each time a VM is created (currently one per transaction processed),
+causing a degradation in runtime performance and likely reducing compute
+by far less compared to the compute savings posed by this proposal.
+
+Even with direct mapping, this change makes address translation more complex
+and therefore slows down _all_ memory accesses.
+
+## New Terminology
+
+N/A.
+
+## Detailed Design
+
+A single `sol_get_sysvar` syscall can be defined, which can be used by all
+forward-facing sysvar interfaces to retrieve data from any sysvar.
+
+```c
+/**
+ * Retrieves a slice of data from a sysvar and copies it to VM memory.
+ *
+ * @param sysvar_id The identifier of the sysvar to retrieve data from.
+ * @param var_addr  VM memory address to copy the retrieved data to.
+ * @param offset    The offset to start copying data from.
+ * @param length    The length of data to copy, in bytes.
+ * @return          A 64-bit unsigned integer error code:
+ *                    - 0 if the operation is successful.
+ *                    - Non-zero error code.
+ *
+ * If the operation is not successful, data will not be written to the
+ * provided VM memory address.
+ */
+uint64_t sol_get_sysvar(
+  /* r1 */ void const * sysvar_id,
+  /* r2 */ uint8_t *    var_addr,
+  /* r3 */ uint64_t     offset,
+  /* r4 */ uint64_t     length,
+);
+
+```
+
+If the following conditions are not met, this syscall will return an error:
+
+- The sysvar data is available to read.
+- The sysvar data is not corrupt.
+- `offset` + `length` is in the range `[0, 2^64)`.
+- `var_addr` + `length` is in the range `[0, 2^64)`.
+- Each byte in the range starting at `var_addr` of size `length` is writable.
```

Comment:
> Sorry, I think I commented this before, but I can't find my comment anymore --
> 
> If the user passes a non-writable range or there is a overflow, the syscall should abort the VM, not return an error. (Which in turn aborts the transaction) This is different from graceful errors, where the syscall just returns non-zero and the VM can continue.

You wrote this on 133, but yes I can update this proposal to say the same thing.

Reply:
I can try to write it a bit more succinctly, but I like the overall approach.

I'm still not sure how you calculate the CU, as I wrote in my reply to you on 133.
Once I understand that better I can update the CU bits.

> - The syscall will always attempt to consume the same amount of CUs irrespective of control flow.

I'm assuming this means that even if we throw an error early on in the execution,
the same number of CUs are consumed as if the process finished successfully?

Reply:
Yeah agreed!

Reply:
@ripatel-fd I think CU calculation should be:

```
syscall_base + (32 / cpi_per_u) + (length / cpi_per_u)
```

Since we're reading the 32-byte sysvar ID and then returning a `length` slice of data. What do you think?

Reply:
Sorry, you seem to agree on the idea in 0127 but not here. These calculations are highly similar. Why use mem_ops.rs?

Reply:
Agave source doesn't do a great job of explaining why these CU values are what they are. For example, why is `syscall_base_cost` and `sysvar_base_cost` equal to `100` while `mem_op_base_cost` is `10`?

I'm assuming there's a reason the sysvar base is much higher than mem_op, so I'm going to use that as a safety net. If anyone explicitly disagrees I'll modify it (@ripatel-fd you said you don't _really_ care).

Thus, the CU calculation I've added is:

```
max(sysvar_base, (32 / cpi_per_u) + (length / cpi_per_u))
```

Reply:
OK here's where we landed, no givesies-backsies.
```
syscall_base + (32 / cpi_per_u) + max(mem_op_base, (length / cpi_per_u))
```